docs/GC Rooting Guide.md

```diff
@@ -0,0 +1,475 @@
+fields, [set the trace
+hook](https://searchfox.org/mozilla-central/search?q=JSTraceOp)
+to invoke a function that traces those fields.
+
+#### General Classes/Structures ####
+
+For a regular `struct` or `class`, tracing must be triggered manually.
+The usual way is to define a
+`void trace(JSTracer* trc, const char* name)` method on the class —
+which is already enough be able to create a `JS::Rooted<YourStruct>` on
+the stack — and then arrange for it to be called during tracing.
+If a pointer to your structure is stored in the private field of a
+JSObject, the usual way would be to define a trace hook on the JSObject
+(see above) that casts the private pointer to your structure and invokes
+`trace()` on it:
+
+```c++
+class MyClass {
+    Heap<JSString*> str;
+
+  public:
+    static void trace(JSTracer* trc, JSObject* obj) {
+        auto* mine = static_cast<MyClass*>(JS_GetPrivate(obj));
+        mine->trace(trc, "MyClass private field");
+    }
+
+    void trace(JSTracer* trc, const char* name) {
+        JS::TraceEdge(trc, &str, "my string");
+    }
+};
+```
+
+If a pointer to your structure is stored in some other structure, then
+its `trace()` method should invoke yours:
+
+```c++
+struct MyOwningStruct {
+    MyClass* mything;
+    void trace(JSTracer* trc, const char* name) {
+        if (mything)
+            mything->trace(trc, "my thing");
+    }
+};
+```
+
+If the toplevel structure is not stored in a `JSObject`, then how it
+gets traced depends on why it should be alive.
+The simplest approach is to use `JS::PersistentRooted` (usable on
+anything with a trace method with the appropriate signature):
+
+```c++
+JS::PersistentRooted<MyOwningStruct> immortalStruct;
+```
+
+But note that `JS::PersistentRooted` in a struct or class is a rather
+dangerous thing to use — it will keep a GC thing alive, and most GC
+things end up keeping their global alive, so if your class/struct is
+reachable in any way from that global, then nothing will ever be cleaned
+up by the GC.
+
+It's also possible to add a custom tracer using
+`JS_AddExtraGCRootsTracer()`.
+Each tracer that gets added needs to be removed again later with
+`JS_RemoveExtraGCRootsTracer()`.
+Obviously it's faster to add (and later remove) one function that gets called during GC and loops over many objects than adding (and removing) many objects to the GC root set.
+
+## Testing Rooting ##
+
+### JS_GC_ZEAL (increased GC frequency) ###
+
+This is a debugging feature to increase the frequency of garbage
+collections.
+It should reveal issues that would only show up in rare cases under
+normal circumstances.
+If the feature is enabled in the SpiderMonkey build (`--enable-gczeal`),
+you can set the environment variable `JS_GC_ZEAL` to configure
+debugging.
+Set it to -1 to print a table of possible settings (or look up that
+table in jsgc.cpp).
+
+The most useful settings probably are:
+
+- 2: GC every F allocations (default: 100)
```

Comment:
JS_GC_ZEAL can also use string names which is probably better advice.

2, Alloc
4, VerifierPre
7, GenerationalGC
11, IncrementalMarkingValidator